crates/common/certificate/src/lib.rs

```diff
@@ -146,6 +148,81 @@ impl KeyCertPair {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum PemCertificateError {
+    #[error("HTTP Connection Problem: {msg} \nHint: {hint}")]
+    CertificateValidationFailure { hint: String, msg: String },
+
+    #[error(transparent)]
+    ReqwestError(#[from] reqwest::Error),
+}
+
+// Our source of error here is quite deep into the dependencies and we need to dig through that to get to our certificates validator errors which are Box<&dyn Error> through 3-4 levels
+// source: hyper::Error(
+//     Connect,
+//     Custom {
+//         kind: Other,
+//         error: Custom {
+//             kind: InvalidData,
+//             error: InvalidCertificateData(
+//                 ..., // This is where we need to get
+//             ),
+//         },
+//     },
+// )
+// At the last layer we have the InvalidCertificateData error which is a Box<&dyn Error> derived from WebpkiError not included anymore, just as a String
+// This chain may break if underlying crates change.
+pub fn get_webpki_error_from_reqwest(err: reqwest::Error) -> PemCertificateError {
+    if let Some(rustls::Error::InvalidCertificateData(inner)) = err
+        // get `hyper::Error::Connect`
+        .source()
+        .and_then(|err| err.source())
+        // From here the errors are converted from std::io::Error.
+        // `Custom` type is `std::io::Error`; this is our first `Custom`.
+        .and_then(|custom_error| custom_error.downcast_ref::<std::io::Error>())
+        .and_then(|custom_error| custom_error.get_ref())
+        // This is our second `Custom`.
+        .and_then(|custom_error2| custom_error2.downcast_ref::<std::io::Error>())
+        .and_then(|custom_error2| custom_error2.get_ref())
+        // Get final error type from `rustls::Error`.
+        .and_then(|rustls_error| rustls_error.downcast_ref::<rustls::Error>())
+    {
+        match inner {
```

Comment:
Here what we need to do to avoid a dependency from the `certificate` crate on `reqwest`:

* Get rid of `PemCertificateError`  and simply add a `CertificateError::CertificateValidationFailure` case.
* Add a function in the `certificate` crate to convert a `rustls::Error::InvalidCertificateData` into a `CertificateError`. This is roughly the code of the `match`.
* Let in the `tedge` crate this `get_webpki_error_from_reqwest` function and the chain `err.source().and_then()...and_then()` to finally call this certificate specific conversion function.

Reply:
Refactored the code as suggested here.